smart-contracts/test/Unlock/migrateLockData.js

```diff
@@ -0,0 +1,368 @@
+  let unlock
+  let lock
+  let pastVersion
+
+  const pastPublicLockPath = require.resolve(
+    `@unlock-protocol/contracts/dist/PublicLock/PublicLockV${versionNumber}.sol`
+  )
+
+  before(async function copyAndBuildContract() {
+    // make sure mocha doesnt time out
+    this.timeout(200000)
+
+    await fs.copy(
+      pastPublicLockPath,
+      path.resolve(contractsPath, `PublicLockV${versionNumber}.sol`)
+    )
+
+    // re-compile contract using hardhat
+    await run('compile')
+  })
+
+  after(async () => {
+    await fs.remove(contractsPath)
+    await fs.remove(artifactsPath)
+  })
+
+  beforeEach(async () => {
+    const [unlockOwner, creator] = await ethers.getSigners()
+
+    // deploy Unlock
+    const Unlock = await ethers.getContractFactory('Unlock')
+    unlock = await upgrades.deployProxy(Unlock, [unlockOwner.address], {
+      initializer: 'initialize(address)',
+    })
+    await unlock.deployed()
+
+    const PublicLockPast = await ethers.getContractFactory(
+      `contracts/past-versions/PublicLockV${versionNumber}.sol:PublicLock`
+    )
+    const PublicLockLatest = await ethers.getContractFactory(
+      'contracts/PublicLock.sol:PublicLock'
+    )
+
+    // deploy past impl
+    const publicLockPast = await PublicLockPast.deploy()
+    await publicLockPast.deployed()
+    pastVersion = await publicLockPast.publicLockVersion()
+
+    // add past impl to Unlock
+    const txImpl = await unlock.addLockTemplate(
+      publicLockPast.address,
+      pastVersion
+    )
+    await txImpl.wait()
+
+    // set v1 as main template
+    await unlock.setLockTemplate(publicLockPast.address)
+
+    // deploy a simple lock
+    const args = [
+      60 * 60 * 24 * 30, // 30 days
+      ethers.constants.AddressZero,
+      ethers.utils.parseEther('0.01'),
+      200, // 200 available keys
+      'A neat upgradeable lock!',
+    ]
+    const calldata = await createLockHash({ args, from: creator.address })
+    const tx = await unlock.createUpgradeableLock(calldata)
+    const { events } = await tx.wait()
+    const evt = events.find((v) => v.event === 'NewLock')
+    const { newLockAddress } = evt.args
+
+    // get lock
+    lock = await ethers.getContractAt(
+      `contracts/past-versions/PublicLockV${versionNumber}.sol:PublicLock`,
+      newLockAddress
+    )
+
+    // deploy latest implementation
+    const publicLockLatest = await PublicLockLatest.deploy()
+    await publicLockLatest.deployed()
+    await unlock.addLockTemplate(publicLockLatest.address, pastVersion + 1)
+  })
+
+  it('lock should have correct version', async () => {
+    assert.equal(await lock.publicLockVersion(), pastVersion)
+  })
+
+  describe('data / schema migration with more than 100 records', () => {
+    let keyOwners
+    let tokenIds
+    let expirationTimestamps
+
+    beforeEach(async () => {
+      const [, generousBuyer] = await ethers.getSigners()
+
+      // create 110 random wallets
+      keyOwners = await Promise.all(
+        Array(110)
```

Comment:
Ok so I think for full testing can we make an even larger number so we make sure that "pagination" works? 
Like 500? ( I assume this will bump into the gas limit?)

Reply:
yep